### PR TITLE
[Community] ResAdapter training script

### LINKED ISSUE #7243: ResAdapter: Domain Consistent Resolution Adapter for Diffusion Models
### Model/Pipeline/Scheduler description

ResAdapter support image generation in arbitrary resolution for comunity models. The result is quite interesting
Project page: https://res-adapter.github.io/
Github: https://github.com/bytedance/res-adapter

### Open source status

- [X] The model implementation is available.
- [X] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

_No response_

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Adds ResAdapter to diffusers community features.

Fixes #7243.

Paper: https://arxiv.org/abs/2403.02084
Code: https://github.com/bytedance/res-adapter
Project page: https://res-adapter.github.io/

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@jiaxiangc @sayakpaul @DN6 @yiyixuxu @rootonchair

### CODE REVIEW

**Comment by a-r-r-o-w** [comment]
@jiaxiangc With lora rank=8, I get around 0.5M trainable parameters which seems to be consistent with the information from the paper. I've followed your suggestions from [here](https://github.com/huggingface/diffusers/issues/7243#issuecomment-1990234653) and believe I've done it correctly, so would be awesome if you could review when free.

I am currently running a training experiment on an A6000 with 1000 training samples but the results do not look good (I'm thinking the dataset is to blame since the split I'm using contains very different images, and lower quality, from what RealisticVision5.1 was finetuned with based on manual inspection. The resolution of images seem to be somewhat limited as well):

<details>
<summary> Script </summary>

```py
#!/usr/bin/bash

python3 train_resadapter.py \
  --pretrained_model_name_or_path krnl/realisticVisionV51_v51VAE \
  --dataset_name poloclub/diffusiondb \
  --dataset_config_name 2m_random_1k \
  --image_column image \
  --caption_column prompt \
  --validation_prompt "beautiful face, youthful appearance, ultra focus, face iluminated, face detailed, ultra focus, dreamlike images, pixel perfect precision, ultra realistic;Award-winning photo of a mystical fox girl fox in a serene forest clearing, sunlight" \
  --validation_prompt_sep ";" \
  --num_validation_images 5 \
  --validation_epochs 1 \
  --validation_heights 256 384 768 768 1024 \
  --validation_widths 256 832 768 1280 1024 \
  --validation_inference_steps 40 \
  --output_dir sd-resadapter \
  --cache_dir . \
  --seed 42 \
  --nearest_resolution_multiple 64 \
  --random_flip \
  --train_batch_size 4 \
  --num_train_epochs 20 \
  --gradient_accumulation_steps 1 \
  --learning_rate 1e-5 \
  --adam_beta1 0.95 \
  --adam_beta2 0.99 \
  --checkpointing_steps 500 \
  --rank 16 \
  --report_to wandb
```

Relevant logs:

```
03/16/2024 19:33:13 - INFO - __main__ - Found 31 unique image sizes
03/16/2024 19:33:13 - INFO - __main__ - Keys: [torch.Size([3, 640, 640]), torch.Size([3, 512, 1024]), torch.Size([3, 512, 512]), torch.Size([3, 704, 512]), torch.Size([3, 512, 768]), torch.Size([3, 1024, 1024]), torch.Size([3, 768, 512]), torch.Size([3, 512, 896]), torch.Size([3, 512, 704]), torch.Size([3, 896, 512]), torch.Size([3, 640, 512]), torch.Size([3, 832, 512]), torch.Size([3, 512, 640]), torch.Size([3, 832, 1024]), torch.Size([3, 576, 1024]), torch.Size([3, 384, 832]), torch.Size([3, 768, 768]), torch.Size([3, 512, 832]), torch.Size([3, 1024, 512]), torch.Size([3, 704, 1024]), torch.Size([3, 512, 960]), torch.Size([3, 512, 1088]), torch.Size([3, 832, 832]), torch.Size([3, 512, 1472]), torch.Size([3, 576, 576]), torch.Size([3, 768, 1024]), torch.Size([3, 1024, 704]), torch.Size([3, 768, 1280]), torch.Size([3, 1024, 768]), torch.Size([3, 1024, 832]), torch.Size([3, 832, 640])]
```
</details>

Training logs: https://wandb.ai/aryanvs/text2image-fine-tune/runs/ogmrnvve

**Comment by rootonchair** [comment]
Hi @a-r-r-o-w , thank you for the good work.
I also think 1000 images is small to achieve good result. I would suggest using a larger dataset like [CC12m](https://github.com/google-research-datasets/conceptual-12m). Besides, have you implemented a probability function to choose different resolution as mentioned by the author in Sec 4.3?

**Comment by a-r-r-o-w** [comment]
Hey. Have not implemented a probability function for selecting ratios as such. See SizeBatchSampler implementation which creates buckets of different image sizes. Prior to creating buckets, images are preprocessed to the nearest resolution multiple of 64 in my training run above, which helps reduce number of buckets. From each bucket, batch size amount of images are selected and once we have all batches, they are randomly shuffled.

Btw, training this on A6000/RTX 4090 takes quite long tbh which makes it hard to test with different datasets. I doubt c12m will have a significant impact here but feel free to try since I have limited compute. The authors train for 20000 steps and single epoch over LAION-5B iirc with batch size 32. I tried a run with 50000 images (49 unique image size buckets) for 2 epochs which was about 4000 steps (and took ~2.5 hours) but the results still didn't seem to be the same as authors on RV5.1. It was definitely starting to get better so I think more images should help.

**Comment by sayakpaul** [comment]
Can we add this to `research_projects` folder to begin with?

**Comment by a-r-r-o-w** [comment]
> Can we add this to `research_projects` folder to begin with?

Sure! Btw, if this, or any research/project in general, is something of interest to the HF team, would it be possible to grant compute for finetuning and creating improved checkpoints? AFAIK diffusers is meant to be an accumulation of research that is widely used within the community and training/finetuning is not as high of a priority as replicating inference behavior is, so I believe the answer is a no, but I'd be really happy to access one of higher end chips (if yes) for replicating open research and work on training scripts as it'd help understand how things work in practice :) The reason I ask is because there are many issues on multiple research repos asking for training code but the requests are simply ignored, possibly due to company policies. I believe there are many that would like to adapt the imaginary training code for their own use cases, or just use it for the sake of openly replicating results for pursuing ideas around improvements.

**Comment by sayakpaul** [comment]
That is also a bit contextual. For high value-add papers (where we have a certain artifact), we can try providing compute grants. 

ADD from Stability AI is a good example of this. They made the model open but didn't probably release the code.

**Comment by jiaxiangc** [comment]
> @jiaxiangc With lora rank=8, I get around 0.5M trainable parameters which seems to be consistent with the information from the paper. I've followed your suggestions from [here](https://github.com/huggingface/diffusers/issues/7243#issuecomment-1990234653) and believe I've done it correctly, so would be awesome if you could review when free.
> 
> I am currently running a training experiment on an A6000 with 1000 training samples but the results do not look good (I'm thinking the dataset is to blame since the split I'm using contains very different images, and lower quality, from what RealisticVision5.1 was finetuned with based on manual inspection. The resolution of images seem to be somewhat limited as well):
> 
> Script
> ```python
> #!/usr/bin/bash
> 
> python3 train_resadapter.py \
>   --pretrained_model_name_or_path krnl/realisticVisionV51_v51VAE \
>   --dataset_name poloclub/diffusiondb \
>   --dataset_config_name 2m_random_1k \
>   --image_column image \
>   --caption_column prompt \
>   --validation_prompt "beautiful face, youthful appearance, ultra focus, face iluminated, face detailed, ultra focus, dreamlike images, pixel perfect precision, ultra realistic;Award-winning photo of a mystical fox girl fox in a serene forest clearing, sunlight" \
>   --validation_prompt_sep ";" \
>   --num_validation_images 5 \
>   --validation_epochs 1 \
>   --validation_heights 256 384 768 768 1024 \
>   --validation_widths 256 832 768 1280 1024 \
>   --validation_inference_steps 40 \
>   --output_dir sd-resadapter \
>   --cache_dir . \
>   --seed 42 \
>   --nearest_resolution_multiple 64 \
>   --random_flip \
>   --train_batch_size 4 \
>   --num_train_epochs 20 \
>   --gradient_accumulation_steps 1 \
>   --learning_rate 1e-5 \
>   --adam_beta1 0.95 \
>   --adam_beta2 0.99 \
>   --checkpointing_steps 500 \
>   --rank 16 \
>   --report_to wandb
> ```
> 
> Relevant logs:
> 
> ```
> 03/16/2024 19:33:13 - INFO - __main__ - Found 31 unique image sizes
> 03/16/2024 19:33:13 - INFO - __main__ - Keys: [torch.Size([3, 640, 640]), torch.Size([3, 512, 1024]), torch.Size([3, 512, 512]), torch.Size([3, 704, 512]), torch.Size([3, 512, 768]), torch.Size([3, 1024, 1024]), torch.Size([3, 768, 512]), torch.Size([3, 512, 896]), torch.Size([3, 512, 704]), torch.Size([3, 896, 512]), torch.Size([3, 640, 512]), torch.Size([3, 832, 512]), torch.Size([3, 512, 640]), torch.Size([3, 832, 1024]), torch.Size([3, 576, 1024]), torch.Size([3, 384, 832]), torch.Size([3, 768, 768]), torch.Size([3, 512, 832]), torch.Size([3, 1024, 512]), torch.Size([3, 704, 1024]), torch.Size([3, 512, 960]), torch.Size([3, 512, 1088]), torch.Size([3, 832, 832]), torch.Size([3, 512, 1472]), torch.Size([3, 576, 576]), torch.Size([3, 768, 1024]), torch.Size([3, 1024, 704]), torch.Size([3, 768, 1280]), torch.Size([3, 1024, 768]), torch.Size([3, 1024, 832]), torch.Size([3, 832, 640])]
> ```
> 
> Training logs: https://wandb.ai/aryanvs/text2image-fine-tune/runs/ogmrnvve

Hi, where can I see your codes? I can give you some details.

**Comment by jiaxiangc** [comment]
@a-r-r-o-w 
Maybe You can open group norm. Finetune on general model, such as SD1.5. 
Make sure your dataset resolutions are right.

Best.

File: examples/resadapter/train_sd_resadapter.py [hunk 1]
```diff
@@ -0,0 +1,1115 @@
...
+from diffusers.utils.torch_utils import is_compiled_module
+
+
+# Will error if the minimal version of diffusers is not installed. Remove at your own risks.
+check_min_version("0.27.0.dev0")
+
+logger = get_logger(__name__, log_level="INFO")
+
+
+class SizeBatchSampler(BatchSampler):
```
Comment (a-r-r-o-w) [inline examples/resadapter/train_sd_resadapter.py:69]: @jiaxiangc Here are the relevant changes I've made to the lora training script for supporting ResAdapter training.

1. SizeBatchSampler allows for the creation of various image size buckets, which are then shuffled randomly. This is equivalent to the probability function that you describe I believe.

File: examples/resadapter/train_sd_resadapter.py [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+
+    # Freeze the unet parameters before adding adapters
+    for param in unet.parameters():
+        param.requires_grad_(False)
+
+    unet_lora_config = LoraConfig(
+        r=args.rank,
+        lora_alpha=args.rank,
+        init_lora_weights="gaussian",
+        target_modules=r".*(up|down)samplers\..*\.conv$",
```
Comment (a-r-r-o-w) [inline examples/resadapter/train_sd_resadapter.py:561]: 2. We add LoRA to up/down-sampler conv layers.

File: examples/resadapter/train_sd_resadapter.py [hunk 3]
```diff
@@ -0,0 +1,1115 @@
...
+    def collate_fn(examples):
+        pixel_values = torch.stack([example["pixel_values"] for example in examples])
+        pixel_values = pixel_values.to(memory_format=torch.contiguous_format).float()
+        input_ids = torch.stack([example["input_ids"] for example in examples])
+        return {"pixel_values": pixel_values, "input_ids": input_ids}
+
+    # DataLoaders creation:
+    train_dataloader = DataLoader(
+        train_dataset,
+        batch_sampler=SizeBatchSampler(train_dataset, shuffle=True, batch_size=args.train_batch_size, drop_last=False),
```
Comment (a-r-r-o-w) [inline examples/resadapter/train_sd_resadapter.py:757]: 3. We use the SizeBatchSampler to only get batches with same image size at a time. Otherwise, torch.stack will fail when collating the tensors.

File: examples/resadapter/train_sd_resadapter.py [hunk 4]
```diff
@@ -0,0 +1,1115 @@
...
+        # Only show the progress bar once on each machine.
+        disable=not accelerator.is_local_main_process,
+    )
+
+    for epoch in range(first_epoch, args.num_train_epochs):
+        unet.train()
+        train_loss = 0.0
+        for step, batch in enumerate(train_dataloader):
+            # freeze group norm if resolution is less than min_groupnorm_unfreeze_resolution
+            if (
```
Comment (a-r-r-o-w) [inline examples/resadapter/train_sd_resadapter.py:846]: 4. We freeze the group norm layers if it is below certain threshold (like less than 512). Otherwise, we make the group norm layers trainable. Note that I'm making both norm1 and norm2 of resnet layer trainable.

**Comment by a-r-r-o-w** [comment]
I think I've been able to somewhat replicate the effect of ResAdapter on my last training run of 50k images. The problem I'm facing is that it is very difficult to train on a large amount of images (> 100k) on a A6000/4090 instance for 1 or 2 epochs. But otherwise, I believe the script is mostly ready and only introduces minimal changes.

**Comment by sayakpaul** [comment]
I think before that it needs to be moved `research_projects` folder?

**Comment by sayakpaul** [comment]
@a-r-r-o-w is this good to be merged?

**Comment by a-r-r-o-w** [comment]
> @a-r-r-o-w is this good to be merged?

@sayakpaul I think it should be good to merge now. However, I would like to note that I haven't fully reproduced the paper results because of limited compute time for this, but the results are clearly noticeable in 1-2 epochs (50k images). One can see structure improve and repetition for large height/width decrease over time. The authors train on much much more data, which is infeasible for me at the moment.

**Comment by a-r-r-o-w** [comment]
I also did a quick training run with 50k images for 1 epoch to verify the latest changes which just completed. The results are for `runwayml/stable-diffusion-v1-5` on first 50k images of LAION-coco-aesthetic.

<table>
<thead>
<tr>
<th>Before training</th>
<th>1'st epoch</th>
</tr>
</thead>
<tbody>
<tr>
<td align=center></td>
<td align=center></td>
</tr>
<tr>
<td align=center></td>
<td align=center></td>
</tr>
</tbody>
</table>



Unfortunately, I can't share training logs or checkpoints due to reasons. Would love to see people from the community pick it up if there's interest.

File: examples/research_projects/resadapter/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+**Code**: https://github.com/bytedance/res-adapter
+**Project Page**: https://res-adapter.github.io/
+
+### Training
+
+```bash
+python3 train_sd_resadapter.py \
+  --pretrained_model_name_or_path <MODEL> \
+  --dataset_name <DATASET> \
+  --dataset_config_name <DATASET_SPLIT> \
```
Comment (sayakpaul) [inline examples/research_projects/resadapter/README.md:15]: Provide a working command here. This means we will need to set `pretrained_model_name_or_path`, `dataset_name` (and other relevant arguments) to their reasonable values.

File: examples/research_projects/resadapter/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+model_id = "<MODEL_ID>"
+resadapter_dir = "<LORA_WEIGHTS_DIRECTORY>"
+pipe = DiffusionPipeline.from_pretrained(
+    model_id,
+    variant="fp16",
+    torch_dtype=torch.float16,
+).to("cuda")
+
+# load resadapter
+unet_groupnorm_state_dict = torch.load(f"{resadapter_dir}/groupnorm_state_dict.pt")
```
Comment (sayakpaul) [inline examples/research_projects/resadapter/README.md:59]: We shouldn't serialize using torch pickle here. `safetensors` is a required dependency of the library so let's use that here. 

Also, can we try to load `f"{resadapter_dir}/groupnorm_state_dict.pt"` from the Hub instead? Something like:


```py
from huggingface_hub import hf_hub_download
import safetensors.torch

state_dict_path = hf_hub_download(repo_id="repo_id", filename="groupnorm_state_dict.pt")
unet_groupnorm_state_dict = safetensors.torch.load_file(state_dict_path)
```

Does this make sense?

File: examples/research_projects/resadapter/README.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    width=768,
+).images[0]
+```
+
+To use the checkpoints provided by the authors, please refer to the original repository linked above.
+
+### Issues
+
+If any problems occur while training/inference, please contact the authors. Two good points of contact would be:
+- [Jiaxiang Cheng](https://github.com/jiaxiangc): First author of the paper.
```
Comment (sayakpaul) [inline examples/research_projects/resadapter/README.md:85]: Maybe first point of contact should be because you implemented the script? And perhaps through a GitHub Discussion on this repository?

**Review by sayakpaul — COMMENTED** [review]
Left some comments.

**Comment by a-r-r-o-w** [comment]
@sayakpaul hey, addressed the previous comments. could you review again? thanks!

**Review by sayakpaul — APPROVED** [review]
Very nice!

**Comment by a-r-r-o-w** [comment]
failing tests seem to be unrelated so far i think
